File: CP(Competetive Programming)/1.Practice/0.Online Virtual Judge/CodingNinja/Merge Sort/MergeSort.py
def mergeSort(arr, n):
    divide(arr, 0, n-1)


def divide(arr, start_point, end_point):
    if (start_point < end_point):
        mid = start_point+int((end_point-start_point)/2)
        divide(arr, start_point, mid)
        divide(arr, mid+1, end_point)
        merge(arr, start_point, mid, end_point)


def merge(arr, start_point, mid, end_point):
    n1 = (mid-start_point)+1
    n2 = end_point-mid
    L = []
    R = []
    for i in range(n1):
        L.append(arr[start_point+i])
    for j in range(n2):
        R.append(arr[mid+1+j])
    i = 0
    j = 0
    k = start_point
    while (i < n1 and j < n2):
        if (L[i] <= R[j]):
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    while (i < n1):
        arr[k] = L[i]
        i += 1
        k += 1
    while (j < n2):
        arr[k] = R[j]
        j += 1
        k += 1


arr = [8, 4, 3, 6]
n = len(arr)
mergeSort(arr, n)

File: CP(Competetive Programming)/1.Practice/0.Online Virtual Judge/CodingNinja/Merge Sort/MergeSort.java
class MergeSort {
    void merge(int[] arr, int startPoint, int mid, int endPoint) {
        int n1 = mid - startPoint + 1;
        int n2 = endPoint - mid;
        int L[] = new int[n1];
        int R[] = new int[n2];
        for (int i = 0; i < n1; i++) {
            L[i] = arr[startPoint + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = arr[mid + 1 + j];
        }
        int i = 0, j = 0;
        int k = startPoint;
        while (i < n1 && j < n2) {
            if (L[i] >= R[j]) {
                arr[k] = R[j];
                j++;
            } else {
                arr[k] = L[i];
                i++;
            }
            k++;
        }
        while (i < n1) {
            arr[k] = L[i];
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    void sort(int[] arr, int startPoint, int endPoint) {
        if (startPoint < endPoint) {
            int mid = startPoint + (endPoint - startPoint) / 2;
            sort(arr, startPoint, mid);
            sort(arr, mid + 1, endPoint);
            merge(arr, startPoint, mid, endPoint);
        }

    }

    public static void main(String[] args) {
        int arr[] = { 8, 4, 3, 6 };
        MergeSort ob = new MergeSort();
        ob.sort(arr, 0, arr.length - 1);
        for (int val : arr) {
            System.out.println(val);
        }
    }
}